### ISSUE TITLE
Refactor .evalDef to be easily extendable

### ISSUE THREAD

**Issue by assignUser** [kgoldfeld/simstudy#13]
# Specifications  for .evalDef

Feel free to add or change anything as you see fit!
The goal is to refactor `.evalDef` to be easily testable, maintainable and extensible.

- [x] test the new .evalDef function thoroughly 

### newVar,defVars:

- [x] [err] newVars is NOT previously defined and 
- [x] [warn] a valid r name -> make.names

### formula:

- [x] [err] dist dependent format checks
- [x] [err] if no defVars: check that formula == scalar

### dist:

- [x] [err] is valid dist

### general arithmetic expression
- [x] [err] all used vars are previously defined or
- [x] are escaped with `..var` and  exist in .GlobalEnv
- [x] [err] is a valid expression build from: numbers, previously defined vars, ~~and base **R**-Operators and functions (e.g. log)~~ valid functions and `..vars`.
- [x] [err] or scalar

### beta
- [x] [err] formula see *g.a.e*
- [x] [err] variance/dispersion value see  *g.a.e*
- [x] [err] link %in% c("identity","logit")

### binary
- [x] [err] ~~formula = [0,1]~~ not posssible until after genData -> *g.a.e*
- [x] [err] link %in% c("identity","logit")
- [x] variance = NULL

### binomial
- [x] [err] ~~formula = [0,1]~~  not posssible until after genData -> *g.a.e*
- [x] [err] ~~variance = integer >= 0~~  not posssible until after genData -> *g.a.e*
- [x] [err] link %in% c("identity","logit")

### categorical
- [x] [err] formula format == "p1;...;pn"
- [x] ~~allow external vars (..var)~~ no real benefit, can just be pasted into formula
- [x] [err] all.equal(sum(probs),1) == TRUE 
- [x] link/variance = NULL   (in the future may be used as category data input)

### exponential 
- [x] [err] formula see *g.a.e*
- [x] [err] link %in% c("identity","log")
- [x] variance = NULL

### gamma 
- [x] [err] formula see *g.a.e*
- [x] [err] variance/dispersion value see  *g.a.e*
- [x] [err] link %in% c("identity","log")

### mixture

- [x] [err] all used vars %in% defVars
- [x] allow external vars (..var)
- [x] [err] all.equal(sum(probs),1) == TRUE 
- [x] [err] formula should have form "x1 | p1 + ... + xn | pn" or "numeric | p1 + ... + numeric | pn"
- [x] link/var = NULL

### negBinomial
- [x] [err] formula see *g.a.e*
- [x] [err] variance/dispersion value see  *g.a.e*
- [x] [err] link %in% c("identity","log")

### nonrandom
- [x] [err] formula see *g.a.e*
- [x] variance/link = NULL

### normal
- [x] [err] formula see *g.a.e*
- [x] [err] variance see  *g.a.e*
- [x] link = NULL

### noZeroPoisson
- [x] [err] formula see *g.a.e*
- [x] [err] link %in% c("identity","log")
- [x] variance = NULL

### poisson
- [x] [err] formula see *g.a.e*
- [x] [err] link %in% c("identity","log")
- [x] variance = NULL

### uniform
- [x] [err] formula format == "min;max"
- [x] ~~allow external vars (..var)~~  no real benefit, can just be pasted into formula
- [x] [err] any(is.numeric(c(min,max)))
- [x] [err] max > min
- [x] [warn] max == min
- [x] link/var = NULL

### uniformInt
- [x] see uniform
- [x] any(is.integer(min,max))

----
Hello!

During the implementation of the `pseudorandom` distribution I came about .evalDef and noticed that it needs to be changed at several places to support a new distribution. It also applies certain checks to all `formula` arguments irrespective of use in the chosen distribution (i had to switch from `"1,2,3;4" to "1+2+3;4"`).

I think the function can be refactored to be easily extendable with new distributions and to allow for more possibilities when parsing arguments.

The function could use templates/patterns/helper functions (naming is hard) to check the applicable    arguments but with tests dependent on the chosen distribution.

**Comment by assignUser** [comment]
Hey @kgoldfeld ,

I'm currently working on this, because while testing tests I noticed a number of cases that were not covered by the function as it is now. 

One of them is, that the mixture formula `1 | 0.5 + 2 | 0.5` or any other where the vars are numbers (which are not valid r varnames!) does not throw an error. This is of course  an easy fix but it got me thinking:  a mixture of nonrandom, scalar variables is just a short cut to have a categorical variable with non-int entries. So we could allow mixtures of the form `numeric | p1 + numeric | p2 + ...` as a shortcut for that. (other checks would still apply ofc). 

Let me know what you think!

**Comment by kgoldfeld** [comment]
I do like the idea of allowing a scalar value instead of a variable there. But, we may also want to have a more explicit way to create these kinds of categories as well - it may be using the "variance" field in the "categorical" distribution to specify the values of the groups, which could be strings or values. We could actually use the same underlying code for the two functions. (One of my biggest regrets is calling the two fields in the data definition table "formula" and "variance" instead of "param1" and "param2", because I've been forced to use the variance field for things that clearly have nothing to do with variance. I would love to change this, but don't think it is possible.)

________________________________

**Comment by assignUser** [comment]
Ok, I will allow scalar values in mixtures and make a note for the new options for categorical data 👍 (I think it is possible to use issues/milestones for stuff like that but I will have to look into that)

Regarding the arguments, thanks to R's lazy evaluation we could do something like this:
```r
par_fun <- function(formula = param1, variance = param2, param1 = formula, param2 = variance) {
  cat(param1,param2)
  cat("\n",formula,variance)
}
 par_fun(formula = "formula", variance = "variance")
#>formula variance
#>formula variance
> par_fun(param1 = "par1",param2 = "par2")
#>par1 par2
#>par1 par2
``` 
Which looks weird and I don't think it is a super good idea? we would need to add additional checks etc.. 

I think the better way is to deal with how it is now and possibly if there are other breaking changes that would allow for more cool features collect them and turn those into simstudy 2.0

**Comment by assignUser** [comment]
# Specifications  for .evalDef

Feel free to add or change anything as you see fit!
The goal is to refactor `.evalDef` to be easily testable, maintainable and extensible.


## General Checks

### newVar,defVars:

- [ ] [err] newVars is NOT previously defined and 
- [ ] [warn] a valid r name -> make.names

### formula:

- [ ] [err] dist dependent format checks
- [ ] [err] if no defVars: check that formula == scalar

### dist:

- [ ] [err] is valid dist

## Distribution Specific Checks

### general arithmetic expression
- [ ] [err] all used vars are previously defined
- [ ] [err] is a valid expression build from: numbers, previously defined vars and base **R**-Operators and functions (e.g. log)
- [ ] [err] or scalar

### beta
- [ ] [err] formula see *g.a.e*
- [ ] [err] variance/dispersion value see  *g.a.e*
- [ ] [err] link %in% c("identity","logit")

### binary
- [ ] [err] formula = [0,1]
- [ ] [err] link %in% c("identity","logit")
- [x] variance = NULL

### binomial
- [ ] [err] formula = [0,1]
- [ ] [err] variance = integer >= 0
- [ ] [err] link %in% c("identity","logit")

### categorical
- [ ] [err] formula format == "p1;...;pn"
- [ ] [err] all.equal(sum(probs),1) == TRUE 
- [x] link/variance = NULL   (in the future may be used as category data input)

### exponential 
- [ ] [err] formula see *g.a.e*
- [ ] [err] link %in% c("identity","log")
- [x] variance = NULL

### gamma 
- [ ] [err] formula see *g.a.e*
- [ ] [err] variance/dispersion value see  *g.a.e*
- [ ] [err] link %in% c("identity","log")

### mixture

- [ ] [err] all used vars %in% defVars
- [ ] [err] all.equal(sum(probs),1) == TRUE 
- [ ] [err] formula should have form "x1 | p1 + ... + xn | pn" or "numeric | p1 + ... + numeric | pn"
- [x] link/var = NULL

### negBinomial
- [ ] [err] formula see *g.a.e*
- [ ] [err] variance/dispersion value see  *g.a.e*
- [ ] [err] link %in% c("identity","log")

### nonrandom
- [ ] [err] formula see *g.a.e*
- [x] variance/link = NULL

### normal
- [ ] [err] formula see *g.a.e*
- [ ] [err] variance see  *g.a.e*
- [x] link = NULL

### noZeroPoisson
- [ ] [err] formula see *g.a.e*
- [ ] [err] link %in% c("identity","log")
- [x] variance = NULL

### poisson
- [ ] [err] formula see *g.a.e*
- [ ] [err] link %in% c("identity","log")
- [x] variance = NULL

### uniform
- [ ] [err] formula format == "min;max"
- [ ] [err] any(is.numeric(c(min,max)))
- [ ] [err] max > min
- [ ] [warn] max == min
- [x] link/var = NULL

### uniformInt
- [ ] see uniform
- [ ] any(is.integer(min,max))

**Comment by assignUser** [comment]
I have finished the refactor, now I will test that everything is as it should be!
You can take a look here: https://github.com/assignUser/simstudy/blob/refactor-evalDef/R/int_evalDef.R

**Comment by kgoldfeld** [comment]
I've meant to follow-up with you. Let me know when we are at a point where we need to make a decisions about how to proceed. I will look at the tests.

**Comment by assignUser** [comment]
closed by #22
